### PR TITLE
8306767: Concurrent repacking of extra data in MethodData is potentially unsafe

### PR DESCRIPTION
**Pull request by eme64**
As explained in a [comment below](https://github.com/openjdk/jdk/pull/16840#issuecomment-1833529561), we have to ensure that reading/writing/cleaning the extra data all needs to be guarded by the `extra_data_lock`, and that no safepoint should happen while holding that lock, so that the lock is not broken.

I introduced `check_extra_data_locked`, where I check that we hold the lock, and if we are a java thread (only those ever safepoint), that we currently are in a `NoSafepointVerifier` scope, hence we verify that no safepoint will be taken.

I placed `check_extra_data_locked` in all the places where we access the extra data, and then placed locks (with implicit no-safepoint-verifiers) at the call-site of those places.

I also needed to change the rank of `extra_data_lock` to `nosafepoint` and set the `Mutex::_no_safepoint_check_flag` when taking the lock. Otherwise I could not take the lock from a VM thread.

**Testing**
Testing: tier1-3 and stress.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8306767](https://bugs.openjdk.org/browse/JDK-8306767): Concurrent repacking of extra data in MethodData is potentially unsafe (**Bug** - P4)

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [ff581b05](https://git.openjdk.org/jdk/pull/16840/files/ff581b0583c02c6b299a972d0f4187322e3282db)
 * [Roland Westrelin](https://openjdk.org/census#roland) (@rwestrel - **Reviewer**) ⚠️ Review applies to [78a2cdb6](https://git.openjdk.org/jdk/pull/16840/files/78a2cdb67c7aa53aba8697f039efe58f68c57ade)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [Tom Rodriguez](https://openjdk.org/census#never) (@tkrodriguez - **Reviewer**) ⚠️ Review applies to [ff581b05](https://git.openjdk.org/jdk/pull/16840/files/ff581b0583c02c6b299a972d0f4187322e3282db)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16840/head:pull/16840` \
`$ git checkout pull/16840`

Update a local copy of the PR: \
`$ git checkout pull/16840` \
`$ git pull https://git.openjdk.org/jdk.git pull/16840/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16840`

View PR using the GUI difftool: \
`$ git pr show -t 16840`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16840.diff">https://git.openjdk.org/jdk/pull/16840.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16840#issuecomment-1829954158)

### CODE REVIEW

**Review by fisk — APPROVED** [review]
Looks good!

File: src/hotspot/share/runtime/deoptimization.cpp [hunk 1]
```diff
@@ -2494,7 +2495,10 @@ Deoptimization::query_update_method_data(MethodData* trap_mdo,
     // Find the profile data for this BCI.  If there isn't one,
     // try to allocate one from the MDO's set of spares.
     // This will let us detect a repeated trap at this point.
-    pdata = trap_mdo->allocate_bci_to_data(trap_bci, reason_is_speculate(reason) ? compiled_method : nullptr);
+    {
+      MutexLocker ml(trap_mdo->extra_data_lock());
```
Comment (tkrodriguez) [inline src/hotspot/share/runtime/deoptimization.cpp:2499]: Doesn't the lock have to be held over the lifetime of the pdata variable?  Otherwise an intervening safepoint could repack the MDO rendering pdata invalid.

Comment (fisk) [reply src/hotspot/share/runtime/deoptimization.cpp:2499]: Good point. That was done in a similar place above but not here.

Comment (eme64) [reply src/hotspot/share/runtime/deoptimization.cpp:2499]: Hmm. So is it that I just have to have a lock on `allocate_bci_to_data`, or do I have to protect `pdata`? Because `pdata` is returned from `Deoptimization::query_update_method_data`, and so I would have to make the lock take a wider scope than that function, right?

Comment (tkrodriguez) [reply src/hotspot/share/runtime/deoptimization.cpp:2499]: The lock needs to be held while accessing the contents of pdata which is very complicated in this particular usage pattern.  Maybe the caller needs to refetch the pointer under a lock instead of passing it out of this method.

Comment (eme64) [reply src/hotspot/share/runtime/deoptimization.cpp:2499]: I see.
And how about other usages of `ProfileData` from the `mdo`, like these (there is a few variants of them around):
```
ProfileData* data = mdo->first_data();
ProfileData* data = mdo->bci_to_data(bci);
```

Comment (eme64) [reply src/hotspot/share/runtime/deoptimization.cpp:2499]: And there are 2 uses of `query_update_method_data`. One does not use the return `pdata`. The other uses it and in some cases updates it. Do you think it is safe to just re-fetch it, or would that potentially cut some connection between the two that should not be cut?
The alternative is just to already get the lock before calling `query_update_method_data`.

Comment (tkrodriguez) [reply src/hotspot/share/runtime/deoptimization.cpp:2499]: I think that anything that can return data from the extra data section is a potential danger.  bci_to_data calls bci_to_extra_data at the end so it seems potentially unsafe which seems like a huge problem since that's used all over the place.  Whether the callers are actually getting or expecting record from extra data is unclear.  I would suspect that most places where it's used there should already be a preallocated record.  The concurrent repacking really makes it hard to ensure the accesses are safe.  I think the API would need to make a stronger split between preallocated records and records which might come from the extra data section.  I'm honestly not sure how to make this truly safe.

Comment (eme64) [reply src/hotspot/share/runtime/deoptimization.cpp:2499]: Ok. So this issue is much bigger than `query_update_method_data` and `allocate_bci_to_data`, is what you are saying. Sounds like I need to study this much deeper. Maybe we need to refactor the while way we access the records? Maybe any access to the records needs to be guarded with a lock, just to be safe?
If there are concurrent updates, which are guarded by lock, then should not all reads also be guarded? In the end maybe we just need to make all accesses mutually exclusive, right?

Comment (fisk) [reply src/hotspot/share/runtime/deoptimization.cpp:2499]: Yeah, reading data from the extra data section, requires holding the lock.

**Comment by eme64** [comment]
@tkrodriguez Yes, we could now do quite the refactoring. The question is how far I should take this. I think I will just make things safe with locks now, and then if someone really wants to fully refactor all of this, then they can do that separately.

**Comment by eme64** [comment]
@fisk @tkrodriguez I continued working on this now, and updated the description. It seems to pass the tests.

One open question is about `MethodData::bci_to_extra_data`:
The idea used to be that we would first read the extra data without a lock (concurrent == true):

`ProfileData* result = bci_to_extra_data_helper(bci, m, dp, true);`

and then if the record was missing, we would take the `extra_data_lock`, and then with `concurrent = false`:

`ProfileData* result = bci_to_extra_data_helper(bci, m, dp, false);`

and then if the record is still missing, we would allocate a new one.

Now, this code is pointless, there is no concurrent action any more because the whole method is under the `extra_data_lock`.

Should I refactor the code there, or leave that to a cleanup-RFE? (for now I left a TODO in the patch).

**Comment by eme64** [comment]
Just fixed an issue from higher tier testing. And I ran performance testing, I think there is no significant difference, so we can go with the simple locking approach that I have now taken.

File: src/hotspot/share/runtime/deoptimization.cpp [hunk 2]
```diff
@@ -2384,6 +2403,15 @@ JRT_ENTRY(void, Deoptimization::uncommon_trap_inner(JavaThread* current, jint tr
       if (make_not_entrant && maybe_prior_recompile && maybe_prior_trap) {
         reprofile = true;
       }
+
```
Comment (rwestrel) [inline src/hotspot/share/runtime/deoptimization.cpp:2406]: Why is it safe to move this here?

File: src/hotspot/share/ci/ciMethodData.cpp [hunk 3]
```diff
@@ -123,12 +130,15 @@ void ciMethodData::prepare_metadata() {
 
 void ciMethodData::load_remaining_extra_data() {
   MethodData* mdo = get_MethodData();
-  MutexLocker ml(mdo->extra_data_lock());
+
+  // Lock to read ProfileData, and ensure lock is not unintentionally broken by a safepoint
+  MutexLocker ml(mdo->extra_data_lock(), Mutex::_no_safepoint_check_flag);
```
Comment (rwestrel) [inline src/hotspot/share/ci/ciMethodData.cpp:135]: Is there anyway to have MutexLocker take care of verifying that the there's no safepoint? It would be nice to replace:

```
MutexLocker ml();
NoSafepointVerifier no_safepoint;
```

by: 

```
MutexLocker ml();
```

only.

File: src/hotspot/share/ci/ciMethodData.cpp [hunk 4]
```diff
@@ -82,13 +82,20 @@ class PrepareExtraDataClosure : public CleanExtraDataClosure {
...
     }
-    // Holding locks through safepoints is bad practice.
-    MutexUnlocker mu(_mdo->extra_data_lock());
+    // We are currently holding the extra_data_lock and ensuring
+    // no safepoint breaks the lock.
+    _mdo->check_extra_data_locked();
+    // We now want to cache some method data. This could cause
+    // a safepoint. We temporarily release the lock and allow
+    // safepoints, and revert that at the end of the scope:
+    MutexUnlocker mu(_mdo->extra_data_lock(), Mutex::_no_safepoint_check_flag);
```
Comment (rwestrel) [inline src/hotspot/share/ci/ciMethodData.cpp:96]: Why is this safe?

File: src/hotspot/share/runtime/deoptimization.cpp (hunk 2 above)
Comment (eme64) [reply src/hotspot/share/runtime/deoptimization.cpp:2406]: Do you think it is not safe?
We have exactly the same conditions:
`make_not_entrant` and `pdata != nullptr` (the second can only hold if we take the current path which gets `pdata`, else it would be `nullptr` anyway).

The only question is with
```
      if (!nm->make_not_entrant()) {
        return; // the call did not change nmethod's state
      }
```

Hmm, maybe I need to copy that here too.But that's kinda ugly. Do you think that is necessary? Any other suggestion?

The idea was to limit the scope of `pdata`. And I kinda have to do that because only in this path do I have a guarantee that `trap_mdo != nullptr`. Otherwise I need to somehow make the lock conditional in the outer scope, that is even more nasty.

File: src/hotspot/share/ci/ciMethodData.cpp (hunk 4 above)
Comment (eme64) [reply src/hotspot/share/ci/ciMethodData.cpp:96]: We only do this in the `finish` method, where we hold no reference to any profiled-data anymore. We only really need to hold the lock during `clean_extra_data` and `is_live`. But after those are done, we can quickly release the lock so that we can call `get_method`.
Does that make sense to you? I'm not super happy with the general pattern here... I basically kept the old pattern.

I wonder, maybe there is a way to move the scope of the lock, such that we only need to lock inside of `clean_extra_data`, and do not hold it before we enter `clean_extra_data`. Do you think that would be preferable?

File: src/hotspot/share/ci/ciMethodData.cpp (hunk 3 above)
Comment (eme64) [reply src/hotspot/share/ci/ciMethodData.cpp:135]: @fisk @tkrodriguez what do you suggest for that?

File: src/hotspot/share/ci/ciMethodData.cpp (hunk 4 above)
Comment (eme64) [reply src/hotspot/share/ci/ciMethodData.cpp:96]: @rwestrel I still believe this is safe. But maybe also ugly.

I looked into making the locking more fine-grained, so that we could avoid unlocking the lock temporarily.
The biggest problem is in `ciMethodData::load_remaining_extra_data`. Here we first (iteratively) clean, and then assume that we still hold the lock when we copy it for the `ciMethodData`. Hence, it seems the lock has to be held at this outer scope, but then temporarily unlocked to allow calls to `get_method` in `PrepareExtraDataClosure::finish`.

Comment (eme64) [reply src/hotspot/share/ci/ciMethodData.cpp:96]: Alternatives to make it prettier:
Make `prepare_metadata` lock, and pass out an object that holds that lock, i.e. widen the scope of the `MutexLocker`. Maybe this can be done with return-value-optimization? But I'm not sure this is a great idea. Another idea @chhagedorn and I thought about was having some Locker object that you can call lock/unlock on, repeatedly. But once the Locker goes out of scope, it checks if it is in the locked state, and only unlocks then. Or maybe it asserts that it is in the locked state, and then unlocks.

Because essencially we need to allow the retry-logic to unlock in between tries. But we also still need to access the `uncached_methods` array that is filled inside the locked region.

I'm not sure such a refactoring is worth it. Let me know what you think @rwestrel

File: src/hotspot/share/ci/ciMethodData.cpp (hunk 3 above)
Comment (eme64) [reply src/hotspot/share/ci/ciMethodData.cpp:135]: Should I create a wrapper object for it? Maybe a `MutexLockerAndNoSafepointVerifier`? Would I place that simply in `methodData.hpp`? Or make it available more widely?

File: src/hotspot/share/runtime/deoptimization.cpp [hunk 5]
```diff
@@ -2312,12 +2327,20 @@ JRT_ENTRY(void, Deoptimization::uncommon_trap_inner(JavaThread* current, jint tr
...
     // aggressive optimization.
     bool inc_recompile_count = false;
+
+    // Lock to read ProfileData, and ensure lock is not broken by a safepoint
+    ConditionalMutexLocker ml((trap_mdo != nullptr) ? trap_mdo->extra_data_lock() : nullptr,
+                              (trap_mdo != nullptr),
+                              Mutex::_no_safepoint_check_flag);
+    NoSafepointVerifier no_safepoint;
     ProfileData* pdata = nullptr;
+
```
Comment (eme64) [inline src/hotspot/share/runtime/deoptimization.cpp:2337]:
```suggestion
```

File: src/hotspot/share/ci/ciMethodData.cpp (hunk 3 above)
Comment (eme64) [reply src/hotspot/share/ci/ciMethodData.cpp:135]: But in this specific instance I keep them separate, because I need to pass the `NoSafepointVerifier` down, so that I can pass it to the `PauseNoSafepointVerifier`.

File: src/hotspot/share/ci/ciMethodData.cpp (hunk 4 above)
Comment (rwestrel) [reply src/hotspot/share/ci/ciMethodData.cpp:96]: Given that logic existed, I think you can leave it as is but a comment that explains why it is safe would be useful.

**Review by rwestrel — APPROVED** [review]
Looks good to me. Thanks for making the chagnes. Do we want to file a bug to refactor the code?

**Comment by eme64** [comment]
@rwestrel thanks for the review! I pushed a comment improvement.

What kind of refactoring are you thinking about? Just to remove the `MutexUnlocker` occurrence, or something bigger?

**Comment by rwestrel** [comment]
> @rwestrel thanks for the review! I pushed a comment improvement.
> 
> What kind of refactoring are you thinking about? Just to remove the `MutexUnlocker` occurrence, or something bigger?

I was thinking maybe something along the lines of what Tom suggested: " I think the API would need to make a stronger split between preallocated records and records which might come from the extra data section. "
It doesn't seem right that the API can return a pointer to something that's not safe to access unless we own a lock too.

**Comment by eme64** [comment]
@rwestrel Ok, makes sense. Would you want to do this?
I mostly addressed this because it is a bug, but otherwise I have my focus elsewhere.

File: src/hotspot/share/runtime/mutexLocker.hpp [hunk 6]
```diff
@@ -254,6 +255,20 @@ class ConditionalMutexLocker: public MutexLockerImpl {
...
+  NoSafepointVerifier _nsv;
+  ConditionalMutexLocker _ml;
+
+public:
+  NoSafepointMutexLocker(Mutex* mutex, bool condition, Mutex::SafepointCheckFlag flag = Mutex::_safepoint_check_flag) :
+    _ml(mutex, condition, flag) {}
+  NoSafepointMutexLocker(Mutex* mutex, Mutex::SafepointCheckFlag flag = Mutex::_safepoint_check_flag) :
+    NoSafepointMutexLocker(mutex, true, flag) {}
+};
+
```
Comment (coleenp) [inline src/hotspot/share/runtime/mutexLocker.hpp:271]: Don't add this.  The locks that are no-safepoint-check-flags have implicit NoSafepointVerifier logic when you take them out.

**Review by coleenp — CHANGES_REQUESTED** [review]
I haven't yet reviewed all of this but this mechanism seems unnecessary and I'd like to understand why this would be added.

File: src/hotspot/share/runtime/mutexLocker.hpp (hunk 6 above)
Comment (eme64) [reply src/hotspot/share/runtime/mutexLocker.hpp:271]: Ah ok. Thanks for the hint. I will look into this. It is a bit confusing what the locks do and do not do with SafePoints. Do they just not safepoint when trying to acquire the lock, or also verify that no safepoint is made while holding the lock?

File: src/hotspot/share/code/compiledMethod.cpp [hunk 7]
```diff
@@ -687,12 +687,15 @@ address CompiledMethod::continuation_for_implicit_exception(address pc, bool for
...
-    print_pcs();
+
+    // Keep tty output consistent. To avoid ttyLocker, we buffer in stream, and print all at once.
+    stringStream ss;
+    ss.print_cr("implicit exception happened at " INTPTR_FORMAT, p2i(pc));
+    print_on(&ss);
+    method()->print_codes_on(&ss);
+    print_code_on(&ss);
+    print_pcs_on(&ss);
+    tty->print("%s", ss.as_string()); // print all at once
```
Comment (coleenp) [inline src/hotspot/share/code/compiledMethod.cpp:698]: It seems like these ttyLocker changes should be checked in as a different cleanup, ie removing ttyLocker is a really good thing.  Can you make these changes a separate patch?

File: src/hotspot/share/runtime/mutexLocker.hpp (hunk 6 above)
Comment (coleenp) [reply src/hotspot/share/runtime/mutexLocker.hpp:271]: They verify that no safepoints happen while holding the lock.  There's a counter in JavaThread.

```
#ifdef ASSERT
  // Debug support for checking if code allows safepoints or not.
  // Safepoints in the VM can happen because of allocation, invoking a VM operation, or blocking on
  // mutex, or blocking on an object synchronizer (Java locking).
  // If _no_safepoint_count is non-zero, then an assertion failure will happen in any of
  // the above cases. The class NoSafepointVerifier is used to set this counter.
  int _no_safepoint_count;                             // If 0, thread allow a safepoint to happen
```

Comment (coleenp) [reply src/hotspot/share/runtime/mutexLocker.hpp:271]: I guess the comment should be updated to say we increment this when taking out a no-safepoint-check mutex.

File: src/hotspot/share/code/compiledMethod.cpp (hunk 7 above)
Comment (eme64) [reply src/hotspot/share/code/compiledMethod.cpp:698]: Yes, I filed the RFE:
[JDK-8324129](https://bugs.openjdk.org/browse/JDK-8324129) C2: Remove some ttyLocker usages in preparation for JDK-8306767
https://github.com/openjdk/jdk/pull/17486

File: src/hotspot/share/runtime/mutexLocker.hpp (hunk 6 above)
Comment (eme64) [reply src/hotspot/share/runtime/mutexLocker.hpp]: Ok, I removed the `NoSafepointVerifier`. It seems to work, and pass my verification that we are in a no_safepoint scope.

**Comment by eme64** [comment]
@coleenp thanks for the suggestions.

I will first integrate this RFE: https://github.com/openjdk/jdk/pull/17486 (please review :wink: )

Then I will integrate and merge it to here.

**Review by fisk — APPROVED** [review]
This looks very promising to me. Very nice with the asserts checking the appropriate locking is there when using the extra data section. Great job!

File: src/hotspot/share/oops/methodData.hpp [hunk 8]
```diff
@@ -2310,7 +2311,9 @@ class MethodData : public Metadata {
   intx arg_local()                               { return _arg_local; }
   intx arg_stack()                               { return _arg_stack; }
   intx arg_returned()                            { return _arg_returned; }
-  uint arg_modified(int a)                       { ArgInfoData *aid = arg_info();
+  uint arg_modified(int a)                       { // Lock and avoid breaking lock with Safepoint
+                                                   MutexLocker ml(extra_data_lock(), Mutex::_no_safepoint_check_flag);
+                                                   ArgInfoData *aid = arg_info();
```
Comment (coleenp) [inline src/hotspot/share/oops/methodData.hpp:2316]: Can you move these to methodData.inline.hpp so that you don't have to include mutexLocker.hpp in a header file?

File: src/hotspot/share/oops/methodData.hpp [hunk 9]
```diff
@@ -2529,6 +2539,18 @@ class MethodData : public Metadata {
...
+    // Cast const away, just to be able to verify the lock
+    // Usually we only want non-const accesses on the lock,
+    // so this here is an exception.
+    MethodData* self = (MethodData*)this;
+    assert(self->extra_data_lock()->owned_by_self(), "must have lock");
+    assert(!Thread::current()->is_Java_thread() ||
+           JavaThread::current()->is_in_no_safepoint_scope(),
+           "JavaThread must have NoSafepointVerifier inside lock scope");
+#endif
+  }
```
Comment (coleenp) [inline src/hotspot/share/oops/methodData.hpp:2553]: Since this is not performance critical, can you move this to the .cpp file?

**Review by coleenp — APPROVED** [review]
This looks better.  I have a couple of small requests.  Thanks.

**Review by tkrodriguez — APPROVED** [review]
This looks good to me.

**Review by coleenp — APPROVED** [review]
Yes, code movement looks good. thank you!
